adam_core/backend/backend.py

```diff
@@ -0,0 +1,168 @@
+
+from ..orbits import Orbits
+from ..utils.indexable import concatenate
+from ..utils.multiprocessing import _check_parallel
+
+logger = logging.getLogger(__name__)
+
+
+def propagation_worker(orbits: Orbits, times: Time, backend: "Backend") -> Orbits:
+    propagated = backend._propagate_orbits(orbits, times)
+    return propagated
+
+
+def ephemeris_worker(orbits: Orbits, observers, backend: "Backend"):
+    ephemeris = backend._generate_ephemeris(orbits, observers)
+    return ephemeris
+
+
+class Backend(ABC):
+    @abstractmethod
+    def _propagate_orbits(self, orbits: Orbits, times: Time) -> Orbits:
+        """
+        Propagate orbits to times.
+
+        THIS FUNCTION SHOULD BE DEFINED BY THE USER.
+        """
+        pass
+
+    def propagate_orbits(
+        self,
+        orbits: Orbits,
+        times: Time,
+        chunk_size: int = 100,
+        num_jobs: int = 1,
+    ) -> Orbits:
+        """
+        Propagate each orbit in orbits to each time in times.
+
+        Parameters
+        ----------
+        orbits : `~adam_core.orbits.orbits.Orbits` (N)
+            Orbits to propagate.
+        times : `~astropy.time.core.Time` (M)
+            Times to which to propagate orbits.
+        chunk_size : int, optional
+            Number of orbits to send to each job.
+        num_jobs : int or "auto", optional
+            Number of jobs to launch. If "auto" then the number of
+            jobs will be equal to the number of cores on the machine.
+
+        Returns
+        -------
+        propagated : `~adam_core.orbits.orbits.Orbits`
+            Propagated orbits.
+        """
+        parallel, num_workers = _check_parallel(num_jobs)
+        if parallel:
+            orbits_split = list(orbits.yield_chunks(chunk_size))
+            times_duplicated = [copy.copy(times) for i in range(len(orbits_split))]
+            backend_duplicated = [copy.copy(self) for i in range(len(orbits_split))]
+
+            p = mp.Pool(
+                processes=num_workers,
+            )
+
+            propagated_list = p.starmap(
+                propagation_worker,
+                zip(
+                    orbits_split,
+                    times_duplicated,
+                    backend_duplicated,
+                ),
+            )
+            p.join()
+            p.close()
+
+            propagated = concatenate(propagated_list)
+        else:
+            propagated = self._propagate_orbits(orbits, times)
+
+        propagated.sort_values(
+            by=["orbit_ids", "times"], ascending=[True, True], inplace=True
+        )
+
+        return propagated
+
+    @abstractmethod
+    def _generate_ephemeris(self, orbits: Orbits, observers):
+        """
+        Generate ephemerides for the given orbits as observed by
+        the observers.
+
+        THIS FUNCTION SHOULD BE DEFINED BY THE USER.
+        """
+        pass
+
+    def generate_ephemeris(
+        self,
+        orbits: Orbits,
+        observers,
```

Comment:
Please annotate the type of `observers`

Reply:
I'm going to hold off on this until I've defined the Observer's class in a follow up PR.